rfc/tap-hardening.md

```diff
@@ -0,0 +1,99 @@
+# Summary
+
+The goal of the Tap Hardening project is to restrict the set of clients that can
+tap resources on the cluster. The current proposal is for the Tap Service to
+become an Extension API server. This will allow requests to be both authorized
+and authenticated through the Kubernetes API. The TapByResource route currently
+served by the public API will be served by the APIService object on the now
+separate Tap Pod.
+
+The data plane proxies will also begin authenticating and authorizing Tap
+requests. The proxy’s Tap server will require a TLS connection and the client
+must be the Tap Service. All other authenticated clients will be rejected.
+
+# Motivation
+
+The Tap feature is open to any user that has access to the cluster. Tap requests
+made through the CLI and the Web Deployment are neither authenticated or
+authorized. The TapByResource route is served by the public API and will start
+Tap connections to any of the resources requested.
+
+When the Linkerd proxies for requested resources receive tap requests, the Tap
+server also does not authenticate or authorize the requesting client. If it can
+open a connection over TLS it will, but it is not required.
+
+# Details
+
+The control plane component of Tap will be referred to as the Tap Controller.
+The data plane component of Tap will be referred to as the Tap server.
+
+## Control Plane
+
+The Tap Service becoming an Extension API server will allow requests to be
+authenticated and authorized through an [aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/).
+The Tap APIService will claim the TapByResource route in the Kubernetes API.
+This will allow the aggregation layer to proxy anything sent to that API path to
+the Tap extension apiserver on the Tap Pod. Tap requests made by the CLI and Web
+Deployment will therefore be proxied through the aggregation layer to the
+APIService.
+
+The Kubernetes API will authenticate the requesting user and authorize the
+rights for the TapByResource API path via RBAC. The Tap extension apiserver will
+authenticate the proxied request from the Kubernetes apiserver and then
+authorize the request from the original user.
```

Comment:
Technically, the tap controller won’t be doing any authz right?

Reply:
I was explicit about this because the documentation leads me to believe the extension apiserver still does some authorization of the request proxied from the Kubernetes apiserver: https://kubernetes.io/docs/tasks/access-kubernetes-api/configure-aggregation-layer/#authentication-flow

> 4. Extension apiserver: authorize the request from the original user

I removed that part, but I can get a better idea of what exactly happens there.

Reply:
In my testing of smi-metrics, the only authz I did was validating the incoming connection from the api server. I was doing some pretty basic RBAC though.